standup/status/search.py

```diff
@@ -0,0 +1,208 @@
+from django.db.models import Q
+
+
+# FIXME(willkg): This ignores the variety of other whitespace characters in unicode.
+WHITESPACE = u' \t\r\n'
+
+
+def to_tokens(text):
+    """Breaks the search text into tokens"""
+    in_quotes = False
+    escape = False
+
+    tokens = []
+    token = []
+
+    for c in text:
+        if c == u'\\':
+            escape = True
+            token.append(c)
+            continue
+
+        if in_quotes:
+            if not escape and c == u'"':
+                in_quotes = False
+            token.append(c)
+
+        elif not escape and c == u'"':
+            in_quotes = True
+            token.append(c)
+
+        elif c in WHITESPACE:
+            if token:
+                tokens.append(u''.join(token))
+                token = []
+
+        else:
+            token.append(c)
+
+        escape = False
+
+    if in_quotes:
+        # Finish off a missing quote
+        if token:
+            token.append(u'"')
+        else:
+            tokens[-1] = tokens[-1] + u'"'
+
+    if token:
+        # Add last token
+        tokens.append(u''.join(token))
+
+    return tokens
+
+
+class ParseError(Exception):
+    pass
+
+
+def unescape(text):
+    """Unescapes text
+
+    >>> unescape(u'abc')
+    u'abc'
+    >>> unescape(u'\\abc')
+    u'abc'
+    >>> unescape(u'\\\\abc')
+    u'\\abc'
+
+    """
+    # Note: We can ditch this and do it in tokenizing if tokenizing
+    # returned typed tokens rather than a list of strings.
+    new_text = []
+    escape = False
+    for c in text:
+        if not escape and c == u'\\':
+            escape = True
+            continue
+
+        new_text.append(c)
+        escape = False
+
+    return u''.join(new_text)
+
+
+def build_match(field, token):
+    return Q(**{'%s__icontains' % field: unescape(token)})
+
+
+def build_match_phrase(field, token):
+    return Q(**{'%s__icontains' % field: unescape(token)})
+
+
+def build_or(clauses):
+    if len(clauses) == 1:
+        return clauses[0]
+
+    q = clauses[0]
+    for clause in clauses[1:]:
+        q = q | clause
+    return q
+
+
+def build_and(clauses):
+    if len(clauses) == 1:
+        return clauses[0]
+
+    q = clauses[0]
+    for clause in clauses[1:]:
+        q = q & clause
+    return q
+
+
+def parse_match(field, tokens):
+    """Parses a match or match_phrase node
+
+    :arg field: the field we're querying on
+    :arg tokens: list of tokens to consume
+
+    :returns: list of match clauses
+    """
+    clauses = []
+
+    while tokens and tokens[-1] not in (u'OR', u'AND'):
+        token = tokens.pop()
+
+        if token.startswith(u'"'):
+            clauses.append(build_match_phrase(field, token[1:-1]))
+        else:
+            clauses.append(build_match(field, token))
+
+    return clauses
```

Comment:
This is similar to the Fjord one, except Fjord is targeting Elasticsearch, so it tokenizes differently.